### ISSUE TITLE
check xcb error again

### ISSUE THREAD

**Issue by martinetd** [swaywm/wlroots#637]
88316e7921942e186df0c6f41fb405fe85bb83f9 (xwm: get rid of xcb call macro) removed most of the _checked() calls, so when there is an error we now get this awesome debug message:

```
2018-02-16 06:11:41 - [xwayland/xwm.c:1015] X11 event: 0
```

[xcb-requests.3](https://www.x.org/archive//X11R7.7/doc/man/man3/xcb-requests.3.xhtml) says event type of 0 means error... So we can cast the event to `xcb_generic_error_t` and get an error code instead, like `3`, which can translate to `window` with a bit of clockwork grease, as well as the opcode of the failed request (as major/minor) which can tell us what failed, like `25`/`0`, which can translate `send_event`, and is the call I messed up to test.

(There's a handy [xcb util-errors](https://cgit.freedesktop.org/xcb/util-errors/) that parses xcb headers at compile time and could help translate these, but it doesn't seem packaged so we probably shouldn't rely on that)

Now, I can't find any way to tell which send_event failed as we use quite a few, but I suppose that'd be a good start for most reports with debug logs -- the question is how to make that decently. As far as I can tell xcb util-errors exists because there is no nice array in xcb headers we could use to programatically tell `send_event` from `25`, we need to manually match against each opcode define like `XCB_SEND_EVENT` until we find one -- there's 120 of these right now even if we don't use all.

What I can do is, add some generic error catching to our event loop, so we get better than what we have; and either print raw values and let folks figure it out or handle one event arbitrarily and print raw values for others and say "write code here if you see that".
Which is better? Any better idea?

**Comment by emersion** [comment]
I think logging the raw error codes is good enough. We can always improve that if we find it inconvenient. But you're right, we should definitely handle error events.

**Comment by SirCmpwn** [comment]
>(There's a handy xcb util-errors that parses xcb headers at compile time and could help translate these, but it doesn't seem packaged so we probably shouldn't rely on that)

This looks cool. If we depend on it distros will package it anyway.

**Comment by psychon** [comment]
Small side note: Just the information that e.g. some `ChangeProperty` request failed might still make debugging a bit complicated if there are many of them and you do not know which one failed. I think I saw some code somewhere (Qt?) which had some `XCB_DEBUG` define. When that is defined, some macro is defined differently so that `__PRETTY_FUNCTION__`, `__FILE__` and `__LINE__` of various places doing xcb requests, plus the sequence number of the request is tracked. This allows additional information of the kind "the error happened somewhere between [here] and [there]" to be printed on errors.

> As far as I can tell xcb util-errors exists because there is no nice array in xcb headers we could use to programatically tell send_event from 25

It's worse than this, extension get assigned some space in the event space, error space and major code space dynamically, i.e. at startup the X11 server assigns them and (in theory) on the next startup it might assign things differently. To figure out the assignment, the code has to explicitly asks for these values (`xcb_query_extension`).
Then, at some point the space for event numbers got almost full and XKEYBOARD had all of its different events have the same event number, but some field in the event tells apart the subevents. Later, the generic event mechanism came around and standardised this (but I think the field is in a different position than what XKEYBOARD does...).

> There's a handy xcb util-errors that parses xcb headers at compile time

Actually, it uses some python code to generate lots of C code containing static tables from XCB's XML description of the protocol, similar to how most of XCB's C code is generated from XML. Parsing xcb headers would be ugly and fragile.

**Comment by martinetd** [comment]
Let's open a follow-up issues for enhancement.
It'd be easy to add a XKB_TRACE(...) macro that would just get the cookie and print the sequence number everytime (along with file and line as you suggest, but the wlr_log functions do that for us).
That's basically what I did manually when I debugged the two errors I noticed at glance, and it really didn't take long to add debug myself as I only had few (<10) calls to trace, but some systematic macroing with optional extra prints could make sense to me.
